pstoremem/addr_book.go

```diff
@@ -96,33 +111,51 @@ func (mab *memoryAddrBook) Close() error {
 // gc garbage collects the in-memory address book.
 func (mab *memoryAddrBook) gc() {
 	now := time.Now()
-	for _, s := range mab.segments {
-		s.Lock()
-		for p, amap := range s.addrs {
+	collect := func(addrs map[peer.ID]map[string]*expiringAddr) []peer.ID {
+		var collectedPeers []peer.ID
+		for p, amap := range addrs {
 			for k, addr := range amap {
 				if addr.ExpiredBy(now) {
 					delete(amap, k)
 				}
 			}
 			if len(amap) == 0 {
-				delete(s.addrs, p)
+				delete(addrs, p)
+				collectedPeers = append(collectedPeers, p)
 			}
 		}
+		return collectedPeers
+	}
+	for _, s := range mab.segments {
+		s.Lock()
+		collect(s.addrs)
+		collected := collect(s.signedAddrs)
+		// remove routing records for peers whose signed addrs have all been removed
+		for _, p := range collected {
+			delete(s.signedPeerRecords, p)
+		}
 		s.Unlock()
 	}
-
 }
 
 func (mab *memoryAddrBook) PeersWithAddrs() peer.IDSlice {
-	var pids peer.IDSlice
+	// deduplicate, since the same peer could have both signed & unsigned addrs
```

Comment:
I think this code has not been visited since we decided to make signed and unsigned records mutually exclusive, no?